### ISSUE TITLE
Improve object navigation

### ISSUE THREAD

**Issue by Adriani90** [nvaccess/nvda#7800]
# Steps to reproduce:
This issue sumarizes the key points discussed in #7759, #7782 and #7749.

# Expected behavior:
Object navigation has a huge potential to provide better accessibility for users in different applications where lots of objects are not included in the tab follow navigation. Apart from that, object navigation brings more productivity since users would not have to remember thousants of keystrokes. A well developed object navigation mode provides the user with the ability to access added functions on ribbons or navigation bars and use them right away while they are working in a certain area of an application without loosing caret position. In many applications, object navigation results in faster reaction of speech feedback. Moreover, object navigation would help to read unlabeled fields or canvas in java applications. A specific context menu which would allow users to label objects would improve individual accessibility. Live regions of applications (e.g. malware bytes while scanning) could easily be followed by object navigation without having to press NVDA + b to read the whole window or to try to find the live regions by moving the physical mouse (which can be quite frustrating for many blind users). Locking this navigation mode similar to browse mode would allow for adding same keystrokes like in browse mode to navigate between objects as described in following approach. Optional window restriction similar to screen review would improve user experience. Optionally overwrite first child to make NVDA remember the navigation path (see attached proof of concept "enhanced navigation 1.0.1" by @Robert-J-H) can improve productivity for the experienced user and standard navigation keys would ease orientation in complex applications. Additionally, reporting of object level and sibling number would significantly ease the practicality of this navigation mode for beginners. Finally, a consistent object structure review implemented in NVDA would solve problems where NVDA gets stuck in a object region like in MS Outlook, Thunderbird or Adobe Reader.

## Concept approach:
1. Adapt technical aspects of screen review in object navigation to work just for the current active window. Add a keystroke for activating and deactivating the restriction
2. see issue #7749. By locking the object navigation you are now able to introduce navigation keystrokes to this navigation mode (e.g. in laptop layout ctrl + page up to go to the very first object level of the window, ctrl + page down to go to the last object level in a window, ctrl + home to go to the first sibling, ctrl + end to go to the last sibling and so on). Say for example we atribute the command NVDA + shift + spacebar to lock and unlock object navigation, the full keystroke for going to the first object level would be NVDA + shift + ctrl + page up. There wouldn't be any conflicts with other keystrokes
3. Add a key stroke for accessing the system menü or the part of the window where you can minimize, maximize or close the window, for sidebars, ribbons etc.
4. Activating object navigation means you are not moving mouse or keyboard cursor. But by pressing m in object navigation would drag the mouse cursor to the current object since the actual keystroke for that is NVDA + shift + m (laptop layout)
5. When object navigation mode is activated, the enter key would simulate a left mouse klick, so to activate an object you press m and then enter
6. Improve mouse draging by NVDA + shift + m (in some applications in object navigation the mouse cursor does not exactly follow the object navigator)
7. Other navigation keys like in browse mode would make the orientation easier (e.g. c for objects which are combo boxes, e for objects which are edit fields, b for objects which are buttons and so on. One advantage would be that you can read the label of a field by pressing left after navigating to a edit field in an application which is actually not very accessible (e.g. Java applications where the label is a canva)
8. Let NVDA remember the navigation path to easily find previous positions (alternative proposal: create a "back command").

# Actual behavior:> Tell us what happens instead.
NVDA's object navigation is still in a prototype stadium. Many users do not use it although it brings lots of advantages. The main reason is that as of now it provides a very confusing user experience. Orientation is highly unconfortable, there are problems in finding the way back to previous objects when navigating out of a window and then back to the previous object (especially when reaching desktop as the first object level).

# System configuration:
## NVDA version:
All

## NVDA Installed or portable:
Both

# Other information:

# Windows version:
all

# Name and version of other software in use when reproducing the issue:
Several applications

# Other questions:
Does the issue still occur after restarting your PC?
Yes

Have you tried any other versions of NVDA?
yes, all

**Comment by Adriani90** [comment]
Please comment in case I forgot any important points.

**Comment by Adriani90** [comment]
I would like to add that there are some users who gave feedback on this navigation mode. But more feedback can be expected if new or improved functions are being implemented. In my opinion, the above aspects bring only positive improvements to object navigation, although some of them would significantly change the way users are currently dealing with it. But it would change it in a posivite manner.

**Comment by Adriani90** [comment]
[EnhancedNavigation-1.0.1-dev.zip](https://github.com/nvaccess/nvda/files/1511332/EnhancedNavigation-1.0.1-dev.zip)

**Comment by feerrenrut** [comment]
@Adriani90 While it's good feed back to see the issues you care about most, this is duplicating existing issues, and will likely cause further confusion about the best place to discuss them. As such I will close this issue.

Some feedback:

Typically, most applications can be interacted with or navigated, without having to use object navigation. Could you elaborate on why you would prefer to use object navigation over focus and browse mode navigation?

> Add a key stroke for accessing the system menü or the part of the window where you can minimize, maximize or close the window, for sidebars, ribbons etc.

Windows has a keystroke `alt+space` to open the system menu for a window (containing minimise, maximise etc). Sidebars / ribbons etc are normally application dependent.

**Comment by feerrenrut** [comment]
I hope this hasn't confused you, I realise this issue may be in response to my comment on #7759:
> a clear summary of the use-cases we have explored in the discussions on this issue, and how they will be addressed would be helpful

The summary I was referring to should be added as a comment on that issue. It should pertain only to the topics discussed on that issue. Ideally highlighting the different ways that people use object navigation. I would suggest this is done in the form of user stories.

**Comment by Adriani90** [comment]
Well I wouldn't care so strong about these issues if they would not be significant in everyday work. I am moderating several groups of NVDA users in Germany and we are discussing these issues before I commit them to github.

**Comment by Adriani90** [comment]
I wouldn't say that I really prefer this navigation mode over the others. It would just be a significant improvement in how we deal with aapplications. In my case and in some other cases that I know, object navigation is used for following:
In Excel to see the formula of a cell whenn the cell is protected, in Browser to reach a notification when f6 is not working (z.B. Firefox), in Java applications to better understand how formfields are labeled and where NVDA gets stuck in tables and tree views and you don't come out by pressing tab or shift tab, in Windows 10 apps where you would have to press 20 times tab to reach the function you would like (sometimes tab or shift tab does not work at all), in malware bytes and other antivirus applications where you have live regions, during the inaccessible installation process of Visual studio 2015 to see which components are activated and which are not, in MS Outlook to simply deal with the calendar, and and and... Unless this mode is not improved and stabilized, you will not get your desired user stories because lots of users don't spend much time in understanding the issues with it. I hope you can understand what it would bring. In any case, no negative effects for the user.

**Comment by feerrenrut** [comment]
Thanks for providing further information, it is helpful to understand the things that get in the way of users.

> It would just be a significant improvement in how we deal with aapplications.

Could you elaborate on this? Is this because of the applications you mention, which have inaccessible elements?

**Comment by Adriani90** [comment]
Hey,

Yes. And not only the mentioned applications. There are also many people who are not very good at remembering where to find some functions in menus. Other people use multiple screen readers and are quite frustrated because they have to learn hundreeds of keystrokes. New keystrokes come with new applications and it is a never ending story for them.

There are other applications like abbyy finereader, where the menu bar is incredibly slugish when using NVEA. So many users prefer to reach their desired functions by object navigation instead of going through the menu bar. You cannot replace menu bars and this should not be a goal. But for many functions object navigation provides a very good workaround.


Best
Adriani
 

Von meinem iPhone gesendet
